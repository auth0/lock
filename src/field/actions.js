import { Map } from 'immutable';
import { swap, updateEntity } from '../store/index';
import { cancelSelection, setField, setOptionField, startSelection } from './index';

export function changeField(id, name, value, validationFn, ...validationExtraArgs) {
  swap(updateEntity, 'lock', id, setField, name, value, validationFn, ...validationExtraArgs);
}

export function startOptionSelection(id, name, iconUrl, icon) {
  // TODO: should be transient
  swap(updateEntity, 'lock', id, m =>
    m
      .setIn(['field', 'selecting', 'name'], name)
      .setIn(['field', 'selecting', 'iconUrl'], iconUrl)
      .setIn(['field', 'selecting', 'icon'], icon)
  );
}

export function selectOption(id, name, option) {
  swap(updateEntity, 'lock', id, m =>
    setOptionField(m.deleteIn(['field', 'selecting']), name, option)
  );
}

export function cancelOptionSelection(id) {
  swap(updateEntity, 'lock', id, m => m.deleteIn(['field', 'selecting']));
}
