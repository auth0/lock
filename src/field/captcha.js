import { setField, setFieldShowInvalid } from './index';

export function validate(captcha) {
  return !!captcha;
}

export function set(m, captcha, wasInvalid) {
  m = setField(m, 'captcha', captcha, validate);
  if (wasInvalid) {
    m = setFieldShowInvalid(m, 'captcha', true);
  }
  return m;
}

export function reset(m, wasInvalid) {
  return set(m, '', wasInvalid);
}
