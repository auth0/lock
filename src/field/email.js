import trim from 'trim';
import { setField } from './index';
import { endsWith } from '../utils/string_utils';
import { isHRDEmailValid } from '../connection/enterprise';
import * as i18n from '../i18n';

const regExp = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

export function validateEmail(str) {
  return isEmail(str);
}

export function isEmail(str) {
  const result = regExp.exec(trim(str.toLowerCase()));
  return !!result && result[0] !== null;
}

export function setEmail(m, str) {
  return setField(m, 'email', str, str => {
    const validHRDEMail = isHRDEmailValid(m, str);

    return {
      valid: validateEmail(str) && validHRDEMail,
      hint: !validHRDEMail ? i18n.html(m, ['error', 'login', 'hrd.not_matching_email']) : undefined
    };
  });
}

export function emailDomain(str) {
  const result = regExp.exec(trim(str.toLowerCase()));
  return result ? result.slice(-2)[0] : '';
}

export function emailLocalPart(str) {
  const domain = emailDomain(str);
  return domain ? str.slice(0, -1 - domain.length) : str;
}
