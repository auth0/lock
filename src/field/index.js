import React from 'react';
import { Map } from 'immutable';
import trim from 'trim';
import OptionSelectionPane from './option_selection_pane';
import * as l from '../core/index';

export function setField(m, field, value, validator = str => trim(str).length > 0, ...args) {
  const prevValue = m.getIn(["field", field, "value"]);
  const prevShowInvalid = m.getIn(["field", field, "showInvalid"], false);
  const validation = validate(validator, value, ...args);

  return m.mergeIn(["field", field], validation, Map({
    value: value,
    showInvalid: prevShowInvalid && prevValue === value
  }));
}

function validate(validator, value, ...args) {
  if (typeof validator != "function") return Map({valid: true});

  const validation = validator(value, ...args);
  return validation && typeof validation === "object"
    ? Map({valid: validation.valid, invalidHint: validation.hint})
    : Map({valid: !!validation});
}

// TODO: this should handle icons, and everything.
// TODO: also there should be a similar fn for regular fields.
export function registerOptionField(m, field, options, initialValue) {
  let valid = true, hasInitial = !initialValue, initialOption;
  options.forEach(x => {
    valid = valid
      && x.get("label") && typeof x.get("label") === "string"
      && x.get("value") && typeof x.get("value") === "string";

    if (!hasInitial && x.get("value") === initialValue) {
      initialOption = x;
      hasInitial = true;
    }
  });

  if (!valid || !options.size) {
    const stopError = new Error(`The options provided for the "${field}" field are invalid, they must have the following format: {label: "non-empty string", value: "non-empty string"} and there has to be at least one option.`);
    stopError.code = "invalid_select_field";
    // TODO: in the future we might want to return the result of the
    // operation along with the model insteand of stopping the
    // rendering, like [false, m] in the case of failure and [true, m]
    // in the case of success.
    return l.stop(m, stopError);
  }

  if (!initialOption) initialOption = Map({});

  return m.mergeIn(["field", field], initialOption, Map({
    options: options,
    showInvalid: false,
    valid: !initialOption.isEmpty()
  }));
}

export function setOptionField(m, field, option) {
  return m.mergeIn(["field", field], option.merge(Map({
    valid: true,
    showInvalid: false
  })));
}

export function isFieldValid(m, field) {
  return m.getIn(["field", field, "valid"]);
}

export function getFieldInvalidHint(m, field) {
  return m.getIn(["field", field, "invalidHint"], "");
}

export function isFieldVisiblyInvalid(m, field) {
  return m.getIn(["field", field, "showInvalid"], false)
    && !m.getIn(["field", field, "valid"]);
}

export function showInvalidField(m, field) {
  return m.setIn(["field", field, "showInvalid"], !isFieldValid(m, field));
}

export function hideInvalidFields(m) {
  return m.update("field", fields => {
    return fields.map(field => field.set("showInvalid", false));
  });
}

// TODO: only used in passwordless, when we update it to use
// validateAndSubmit this won't be needed anymore.
export function setFieldShowInvalid(m, field, value) {
  return m.setIn(["field", field, "showInvalid"], value);
}

export function clearFields(m, fields) {
  let keyPaths;

  if (!fields || fields.length === 0) {
    keyPaths = ["field"];
  } else {
    keyPaths = fields.map(x => ["field", x]);
  }

 return keyPaths.reduce((r, v) => r.removeIn(v), m);
}

export function getField(m, field, notFound=new Map({})) {
  return m.getIn(["field", field], notFound);
}

export function getFieldValue(m, field, notFound="") {
  return getField(m, field).get("value", notFound);
}

export function getFieldLabel(m, field, notFound="") {
  return getField(m, field).get("label", notFound);
}

// phone number

export function phoneNumber(lock) {
  return lock.getIn(["field", "phoneNumber", "value"], "");
}

// email

export function email(m) {
  return getFieldValue(m, "email");
}

// vcode

export function vcode(m) {
  return getFieldValue(m, "vcode");
}

// password

export function password(m) {
  return getFieldValue(m, "password");
}

// username

export function username(m) {
  return getFieldValue(m, "username");
}

// mfa_code

export function mfaCode(m) {
  return getFieldValue(m, "mfa_code");
}

// select field options

export function isSelecting(m) {
  return !!m.getIn(["field", "selecting"]);
}

export function renderOptionSelection(m) {
  const name = m.getIn(["field", "selecting", "name"]);
  return isSelecting(m)
    ? <OptionSelectionPane
         model={m}
         name={name}
         icon={m.getIn(["field", "selecting", "icon"])}
         iconUrl={m.getIn(["field", "selecting", "iconUrl"])}
         items={m.getIn(["field", name, "options"])}
      />
    : null;
}
