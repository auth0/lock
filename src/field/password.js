import PasswordPolicy from 'password-sheriff/lib/policy';
import { setField } from './index';

export function validatePassword(password, policy) {
  if (!policy) {
    return true;
  }
  return new PasswordPolicy(policy.toJS()).check(password);
}

export function setPassword(m, password, policy) {
  return setField(m, 'password', password, validatePassword, policy);
}

export function setShowPassword(m, checked) {
  return setField(m, 'showPassword', checked, () => true);
}
