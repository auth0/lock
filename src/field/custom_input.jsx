import React from 'react';
import { changeField, startOptionSelection } from './actions';
import {
  getFieldInvalidHint,
  getFieldLabel,
  getFieldValue,
  isFieldVisiblyInvalid
} from './index';
import TextInput from '../ui/input/text_input';
import SelectInput from '../ui/input/select_input';
import CheckboxInput from '../ui/input/checkbox_input';
import * as l from '../core/index';

const CustomInput = ({iconUrl, model, name, placeholder, type, validator}) => {
  const props = {
    iconUrl,
    isValid: !isFieldVisiblyInvalid(model, name),
    name,
    placeholder
  };

  switch(type) {
    case "select":
      return (
        <SelectInput
          {...props}
          label={getFieldLabel(model, name)}
          onClick={() => startOptionSelection(l.id(model), name, iconUrl)}
        />
      );
    case "checkbox":
      return (
          <CheckboxInput
              onChange={e => changeField(l.id(model), name, `${e.target.checked}`, validator)}
              checked={getFieldValue(model, name)}
              {...props}
        />
      );
    default:
      return (
        <TextInput
          invalidHint={getFieldInvalidHint(model, name)}
          onChange={e => changeField(l.id(model), name, e.target.value, validator)}
          value={getFieldValue(model, name)}
          {...props}
        />
      );
  }
};

export default CustomInput;
