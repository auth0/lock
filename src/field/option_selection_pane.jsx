import React from 'react';
import List from '../ui/list';
import { cancelOptionSelection, selectOption } from './actions';
// TODO: these actions should be passed as props

export default ({ icon, iconUrl, model, name, items }) => (
  <List
    icon={icon}
    iconUrl={iconUrl}
    items={items}
    onSelect={x => selectOption(model.get('id'), name, x)}
    onCancel={() => cancelOptionSelection(model.get('id'))}
  />
);
