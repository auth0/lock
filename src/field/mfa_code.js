import { setField } from './index';
import { validateEmail } from './email';
import { databaseConnection } from '../connection/database';
import trim from 'trim';

const DEFAULT_VALIDATION = { mfa_code: { length: 6 } };
const regExp = /^[0-9]+$/;

function validateMFACode(str, settings = DEFAULT_VALIDATION.mfa_code) {
  const value = trim(str);

  // check min value matched
  if (value.length < settings.length) {
    return false;
  }

  // check max value matched
  if (value.length > settings.length) {
    return false;
  }

  // check allowed characters matched
  const result = regExp.exec(value);
  return result && result[0];
}

export function setMFACode(m, str) {
  return setField(m, 'mfa_code', str, validateMFACode);
}

export function getMFACodeValidation(m) {
  return DEFAULT_VALIDATION.mfa_code;
}
