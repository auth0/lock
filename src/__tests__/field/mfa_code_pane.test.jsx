import React from 'react';
import { mount } from 'enzyme';

import { expectComponent, extractPropsFromWrapper, mockComponent } from 'testUtils';

jest.mock('ui/input/mfa_code_input', () => mockComponent('mfa_code_input'));

const getComponent = () => require('field/mfa-code/mfa_code_pane').default;

describe('MFACodePane', () => {
  const defaultProps = {
    i18n: {
      str: (...keys) => keys.join(',')
    },
    lock: {},
    placeholder: 'placeholder'
  };

  beforeEach(() => {
    jest.resetModules();

    jest.mock('field/index', () => ({
      getFieldValue: () => 'mfa',
      isFieldVisiblyInvalid: () => true
    }));

    jest.mock('field/password', () => ({
      getMFACodeValidation: () => 'getMFACodeValidation',
      setMFACode: 'setMFACode'
    }));
    
    jest.mock('core/index', () => ({
      id: () => 1,
      submitting: () => false,
      ui: {
        avatar: () => false
      }
    }));

    jest.mock('store/index', () => ({
      swap: jest.fn(),
      updateEntity: 'updateEntity'
    }));
  });

  it('renders correctly', () => {
    const MFACodePane = getComponent();
    expectComponent(
      <MFACodePane
        {...defaultProps}
        />
    ).toMatchSnapshot();
  });
  it('sets isValid as true when `isFieldVisiblyInvalid` is false', () => {
    require('field/index').isFieldVisiblyInvalid = () => false;
    let MFACodePane = getComponent();

    expectComponent(
      <MFACodePane
        {...defaultProps}
        />
    ).toMatchSnapshot();
  });
  it('calls `swap` onChange', () => {
    let MFACodePane = getComponent();

    const wrapper = mount(<MFACodePane {...defaultProps} />);
    const props = extractPropsFromWrapper(wrapper)
    props.onChange({ target: { value: 'newUser' } });

    const {mock} = require('store/index').swap;
    expect(mock.calls.length).toBe(1);
    expect(mock.calls[0]).toMatchSnapshot();
  });
});
