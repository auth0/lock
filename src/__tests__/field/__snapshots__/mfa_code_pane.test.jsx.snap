// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MFACodePane calls \`swap\` onChange 1`] = `
Array [
  "updateEntity",
  "lock",
  1,
  [Function],
  "newUser",
]
`;

exports[`MFACodePane renders correctly 1`] = `
<div
  data-__type="mfa_code_input"
  data-invalidHint="mfaCodeErrorHint,6"
  data-isValid={false}
  data-lockId={1}
  data-onChange={[Function]}
  data-placeholder="placeholder"
  data-value="mfa"
/>
`;

exports[`MFACodePane sets isValid as true when \`isFieldVisiblyInvalid\` is false 1`] = `
<div
  data-__type="mfa_code_input"
  data-invalidHint="mfaCodeErrorHint,6"
  data-isValid={true}
  data-lockId={1}
  data-onChange={[Function]}
  data-placeholder="placeholder"
  data-value="mfa"
/>
`;
